koswat/cost_report/io/summary/summary_infrastructure_costs/summary_infrastructure_costs_csv_fom_builder.py

```diff
@@ -0,0 +1,201 @@
+                    )
+                )
+            return _totals
+
+        def _get_details(location: PointSurroundings, profile_type: str) -> list[float]:
+            _details = []
+            for _infra_type in self._ordered_infra_types:
+                _ilc = _infra_cost_per_loc_dict[location][profile_type][_infra_type]
+                _details.extend(
+                    [_ilc.zone_a, _ilc.zone_a_costs, _ilc.zone_b, _ilc.zone_b_costs]
+                )
+            return _details
+
+        def _location_as_row(
+            matrix_item: tuple[PointSurroundings, list[int]]
+        ) -> list[Any]:
+            _ps, _m_values = matrix_item
+            _location_as_row = [_ps.section, _ps.location.x, _ps.location.y]
+            _location_as_row.extend(_m_values)
+            return _location_as_row
+
+        _infra_cost_per_loc_dict = self._get_infra_location_costs(
+            locations_profile_report_list
+        )
+
+        # Initiate locations matrix
+        _matrix = defaultdict(list)
+
+        # Totals and details per location for profile types
+        for _rpl in reinforcement_per_locations:
+            _totals = _get_totals(_rpl.location)
+            _details = []
+            for _profile_type in self._ordered_profile_types:
+                _details.extend(_get_details(_rpl.location, _profile_type))
+            _matrix[_rpl.location] = _totals + _details
+
+        return list(
+            map(
+                _location_as_row,
+                sorted(_matrix.items(), key=lambda x: x[0].traject_order),
+            )
+        )
+
+    def _get_infra_location_costs(
+        self,
+        locations_profile_report_list: list[MultiLocationProfileCostReport],
+    ) -> dict[PointSurroundings, dict[str, dict[str, InfrastructureLocationCosts]]]:
+        """
+        Create a dictionary with the infrastructure cost per location/profile type/infra type.
+
+        Args:
+            locations_profile_report_list (list[MultiLocationProfileCostReport]):
+                List containing all the chosen profiles with their costs.
+
+        Returns:
+            dict[ PointSurroundings, dict[str, dict[str, InfrastructureLocationProfileCostReport]] ]:
+                Dict containing the infrastructure cost per location/profile type/infra type.
+        """
+        _infra_location_cost_dict: dict[
+            PointSurroundings,
+            dict[str, dict[str, InfrastructureLocationCosts]],
+        ] = defaultdict(
+            lambda: defaultdict(lambda: defaultdict(InfrastructureLocationCosts))
+        )
+        for _lpr in locations_profile_report_list:
+            for _ilpcr in _lpr.infra_multilocation_profile_cost_report:
+                _infra_location_cost_dict[
+                    _ilpcr.infrastructure_location_costs.location
+                ][_lpr.profile_type_name][
+                    _ilpcr.infrastructure.infrastructure_name
+                ] = _ilpcr.infrastructure_location_costs
```

Comment:
Perhaps it's the typography, but this became unreadable to me.
```suggestion
        def update_infra_location_cost_dict(profile_type_name: str, location: Point, infra_name: str, costs: float):
            _infra_location_cost_dict[location][profile_type_name][infra_name] = costs

        for _lpr in locations_profile_report_list:
            for _ilpcr in _lpr.infra_multilocation_profile_cost_report:
                update_infra_location_cost_dict(
                     _lpr,
                     _ilpcr.infrastructure_location_costs.location,
                     _ilpcr.infrastructure.infrastructure_name,
                     _ilpcr.infrastructure_location_costs
                )
```

Reply:
Better!